include/envoy/server/listener_manager.h

```diff
@@ -168,6 +168,11 @@ class ListenerManager {
    * have exited.
    */
   virtual void stopWorkers() PURE;
+
+  /*
+   * Warn the listener of an impending update. This allows the listener to clear per-update state.
+   */
+  virtual void beginListenerUpdate() PURE;
```

Comment:
How will this work with incremental LDS? In that world, we need to selectively clear only those listeners that appear in an update.

Reply:
I was going to comment on this also but forgot. I'm wondering why each listener can't just handle/clear its state independently? Or is the idea that a config might fail to parse/verify before we call into the LM? In that case I'm wondering per @htuch if we should just have a per-listener clear method? That seems a bit less fragile to me.

Reply:
My concern here is making sure we don't leak non-named listeners, where we generate a unique uuid in addOrUpdateListener.    I brought this up on the community call and said I could either clear after a configured time interval, or during the next update, and you both suggested clearing on the next update.  I suppose for incremental LDS we'd have to switch to timer based?

Reply:
For clarity, my concern is that I would like to expose the details for listening failure even for listeners which are never instantiated, due to configuration load failure.  Obviously if we create a listener we clear the state on any LDS update which removes the listener, and for incremental LDS updates we can clear error state on addition / update.  But I still want to expose failure of listeners which never load, and we don't want to leak those.

Reply:
Yeah, this makes sense. I wonder if there is any possible overlap with https://github.com/envoyproxy/envoy/issues/7868, i.e. if a TTL of some sort might apply not only to a resource like the Listener's config, but also its config dump representation.

Reply:
So are we OK with this as-is?   Do you want me to switch to a timer now or just dealing with security backlog and you'll review the rest?

Reply:
Ah OK, makes sense re: UUIDs. However, I think we have a larger problem which is that things can fail before we get here?

https://github.com/envoyproxy/envoy/blob/2cd917e12e218f569773a6b7bdef8c2d72cdc897/source/server/lds_api.cc#L56-L61

I think we need to somehow reconcile all of these possible failure paths so you capture things like duplicates, validation errors, etc. also?

Reply:
Ah yeah, the initial version with the status manger handled global errors by not trying to associate errors with specific listeners, just tracking overall status

I guess the best we can do when trying to tie it to the config dump is to have one more UpdateFailureState tied to the parent ListenersConfigDump, which carries top level config (potentially all listeners) and error messages (whatever was thrown), and again for now it'd be cleared on update but we make it repeated so that eventually when we have incrementalLds we can clear it on a timer basis.  SG?

Reply:
Yup SGTM

Reply:
Re: a parent failure status, +1. I'd point out though, that in the above example, we can probably narrow down both the duplicate and some of the validation errors by performing a per-listener check first. I'd just leave a TODO there, but it's an opportunity for more precision down the line.

Re: timer, yeah, let's just do it now. I think having a dedicated TTL for failed listener updates is totally fine, the TTL can be long. TBH, the trend I'm hearing from the folks who are using LDS (e.g. Istio, TD) is that they want better inline error messages, this is really just there for humans, so I'm not too concerned about losing information in monitoring etc.

Reply:
Tracking as https://github.com/envoyproxy/envoy/issues/8618 as discussed offline